### ISSUE TITLE
Add ImmutableGeoPoint Class

### ISSUE THREAD

**Issue by nknize** [elastic/elasticsearch#13892]
The current GeoPoint class provides `reset` utility methods making all GeoPoint instance volatile. This is particularly dangerous for existing Query objects but necessary for Grid and Centroid metric aggregations (situations that would otherwise create many transient GeoPoint objects). This enhancement adds an ImmutableGeoPoint class for guarding against potential query corruption.

**Comment by rjernst** [comment]
Can we just make GeoPoint immutable without adding another class?

**Comment by nknize** [comment]
After thinking a bit more, how about a simple mutable GeoPoint class that extends an immutable Point base class? Since we want to eventually support 3 or 4 dimensions this would give us both the read-only protection we need for Query state and the foundation for extending into higher dimensions - without creating a class for each dimension like other libraries do (e.g., Point2D, Point3D).  This would be very useful for coordinate reprojection and 3rd party plugin support.

**Comment by danielmitterdorfer** [comment]
Although it seems practical, this breaks the Liskov Substitution Principle and can lead to surprises:

```
//MutablePoint inherits from ImmutablePoint
MutablePoint pm = new MutablePoint(0, 0);
ImmutablePoint pi = pm;

// in another method far, far away, the x coordinate is cached since pi is immutable

final int x = pi.getX();

// now we update the value somewhere else again ...
pm.setX(3);

// ... and the cached value 'x' is wrong :(
```

It is also tricky to implement concurrent updates correctly. I think you would have to reimplement the mutable point class from scratch anyway and also provide an API that allows for atomic updates of a point coordinate (assuming concurrent updates are supported).

I am also not sure how your design would help extending to more than two dimensions.

**Comment by nknize** [comment]
This is how I was planning: https://github.com/nknize/elasticsearch/blob/immutableGeopoint/core/src/main/java/org/elasticsearch/common/geo/GeoPoint.java

But this may be OBE as I believe we can make GeoPoint immutable (as Ryan mentioned) with low impact and less risk.

**Comment by danielmitterdorfer** [comment]
Thanks for the code example. This makes it much clearer.

However, as the `GeoPoint` fields are not final, the class is not immutable but just effectively immutable. This will be no problem as long as `GeoPoint` is only used from a single thread but you (might) get into trouble if you have multiple threads and you publish `GeoPoint` with a data race. If they fields of `GeoPoint` would be final, you'd be even safe in this case (due to the guarantees defined in [JLS 17.5](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5)).

Long story short, this would work perfectly fine in a single threaded case but it is not safe for multithreading.

**Comment by nknize** [comment]
Thanks for the review and great input @danielmitterdorfer  If/when this lands I'll give you a heads up to review the PR. @colings86 how urgent is GeoPoint immutability?

**Comment by colings86** [comment]
@nknize its not that urgent. More like a nice to have really. It just means we won't need to do any copying of points to make sure the don't change

**Comment by danielmitterdorfer** [comment]
@nknize Thanks, I'm glad I could help. :)

**Comment by javanna** [comment]
@nknize Is this something that has yet to be done? Anything left to discuss on the matter?

**Comment by nknize** [comment]
closing for now.. it hasn't been that pressing and we can reopen later.
